Data Structures. First commit for data structures in Ruby.

sungchoi opened the pull request:
Completely to mostly finished data structures:
Linked List
Doubly Linked List
Stack
MinStack
MaxStack
MinmaxStack
Queue
Binary Tree

sungchoi committed:
Data Structures. First commit for data structures in Ruby.
Completely to mostly finished data structures:
Linked List
Doubly Linked List
Stack
MinStack
MaxStack
MinmaxStack
Queue
Binary Tree

sungchoi committed:
Data Structures. Doubly Linked List. Changes length to be 0 by default.
Doubly Linked List does not yet support adding a collection at initialization or afterwards.

sungchoi committed:
Data Structures. Ruby. Linked List. Doubly Linked List.
Adds stubs for #unshift_all(enumerable).

sungchoi committed:
Data Structures. Ruby. Updates Gemfile by adding rspec.

sungchoi committed:
Data Structures. Ruby. Updates file paths in queue_spec.rb and stack_spec.rb

sungchoi committed:
Data Structures. Ruby. Updates BinaryTree::Node #each and related methods to properly pass blocks and to check for existence of left and right before making method calls on them.

sungchoi committed:
Data Structures. Ruby. DoublyLinkedList. Makes tail, head, and length read only (attr_reader).

jfarmer commented on data-structures/ruby/binary_tree/binary_tree_node.rb:
```diff
@@ -0,0 +1,12 @@
+module BinaryTree
+
+  def BinaryTree(value, left = nil, right = nil)
```
I would expect this to be in the same file as the `BinaryTree` class.

jfarmer commented on data-structures/ruby/doubly_linked_list/base.rb:
```diff
@@ -0,0 +1,123 @@
+module DoublyLinkedList
+  class Base
+
+    include DoublyLinkedList
```
I'm not sure where/why you picked up this pattern.  You're module-crazy!

jfarmer commented on data-structures/ruby/doubly_linked_list/base.rb:
```diff
@@ -0,0 +1,123 @@
+
+    def unshift(value)
+      node = DoublyLinkedListNode(value, @head)
+      if @head.is_a?(Empty)
```
Type checking is a code smell.  Make `Node` and `EmptyNode` implement the same interface so that things that use them need not care which type the object is.

jfarmer commented on data-structures/ruby/doubly_linked_list/base.rb:
```diff
@@ -0,0 +1,123 @@
+    end
+
+    def remove_from_tail
+      return UnderflowError.new("Underflow Error: Linked List is empty") if empty?
```
"Underflow" isn't generally used in the context of linked lists.  When taking about data structures, it's a stack-specific phrase.  Sometimes we use the phrase _arithmetic underflow_ to describe a situation where we're trying to deal with floating point numbers smaller than a computer can represent.

jfarmer commented on data-structures/ruby/doubly_linked_list/base.rb:
```diff
@@ -0,0 +1,123 @@
+    end
+
+    def empty?
+      @head.is_a?(DoublyLinkedList::Empty)
```
Why not just define an `empty?` method on the nodes?  It can return true for the default type and false for the empty type.

jfarmer commented on data-structures/ruby/binary_tree/node.rb:
```diff
@@ -0,0 +1,55 @@
+
+    # @return [Array<BinaryTree::Node>]
+    def children
+      [@left, @right]
```
Use the public interface when possible, i.e.,

``` ruby
def children
  [left, right]
end
```

jfarmer commented on data-structures/ruby/doubly_linked_list/empty.rb:
```diff
@@ -0,0 +1,11 @@
+module DoublyLinkedList
+  class Empty
```
Isn't `Empty` a refined type of `Node`?  That should push you in the direction of making `Empty` a subclass of `Node`.

jfarmer commented on data-structures/ruby/min_stack.rb:
```diff
@@ -0,0 +1,24 @@
+  def pop
+    @max_stack.pop if peek == @max_stack.peek
+    super
+  end
```
Hmm, this works — I think!  Consider using a single `Stack` and pushing a tuple onto it instead of a single value.  The first element of the tuple would be the value and the second element would be the largest thing in the stack at that point.

``` ruby
def push(value)
  super([value, [value, max].max])
end
```

You might need to rejigger it a bit to get all the methods to work, but that's the main idea.

jfarmer commented on data-structures/ruby/queue.rb:
```diff
@@ -0,0 +1,35 @@
+  end
+
+  def initialize()
+    @list = DoublyLinkedList::Base.new()
```
We don't actually need a doubly linked list to implement a queue.  Remember, although we talk about queues as if there's a "front" or an "end", really what we care about is the fact that we can put information into the queue and later retrieve the oldest information in the queue.

If we're using some kind of list to power these operations, all that means is we need to be able to add elements to one end in constant time and remove them from the other in constant time.  It doesn't matter which end of the list, i.e., there's no reason for us to require that the "front" of the queue corresponds to the "front" of the list.

Singly linked lists always have constant time addition/removal from the head.  We can augment it with a constant time append operation by keeping track of the tail.  Enqueuing would then map to appending information to the end of the list and dequeuing would map to removing information from the head of the list.

jfarmer commented on data-structures/ruby/spec/binary_tree/node_spec.rb:
```diff
@@ -0,0 +1,46 @@
+    #TODO
+    expect(tree.left == BinaryTree::Node.new(1)).to eq(true)
+    expect(tree.right == BinaryTree::Node.new(3)).to eq(true)
+    expect(tree == BinaryTree::Node.new(2)).to eq(false)
```
Why not this?

``` ruby
expect(tree.left).to eq(BinaryTree::Node.new(1))
expect(tree).not_to eq(BinaryTree::Node.new(2))
```

jfarmer commented on data-structures/ruby/stack.rb:
```diff
@@ -0,0 +1,54 @@
+  # @time BigO(1)
+  def pop
+    if @list.length == 0
+      UnderflowError.new("Underflow Error: stack is empty")
```
It's weird to _return_ an exception rather than _raise_ one.

jfarmer commented on data-structures/ruby/stack.rb:
```diff
@@ -0,0 +1,54 @@
+    @list == other.list
+  end
+
+end
```
I added a spec for stack to the repo itself, if you want to compare: https://github.com/codeunion/data-structures/blob/master/spec/stack_spec.rb

jfarmer commented on data-structures/ruby/doubly_linked_list/base.rb:
```diff
@@ -0,0 +1,123 @@
+      @tail
+    end
+
+    def calculate_length
```
Why not call this `length`?

jfarmer commented on data-structures/ruby/doubly_linked_list/base.rb:
```diff
@@ -0,0 +1,123 @@
+      length
+    end
+
+    def find_tail
```
What is the point of this method?  Aren't you keeping track of the tail directly?

jfarmer commented on data-structures/ruby/linked_list/node.rb:
```diff
@@ -0,0 +1,29 @@
+    end
+
+    def insert_before(value)
+      node = LinkedListNode(value)
```
Why not this?

``` ruby
def insert_before(value)
  LinkedListNode(value, self.next)
end
```

sungchoi committed:
Data Structures. Ruby. BinaryTree. Reorganizes namespacing.
Deletes BinaryTree module. Creates BinaryTreeNode class instead.

sungchoi committed:
Data Structures. Ruby. Tree. Reorganizes Tree.
Removes Tree module and creates TreeNode class.

sungchoi committed:
Data Structures. Ruby. DoublyLinkedList. Changes how the module handles empty?
Empty inherits from Node. Empty#empty? returns true. Node#empty? returns false.
Removes type checks for Empty using #empty? instead.

sungchoi committed:
Data Structures. Ruby. DoublyLinkedList. Makes @prev and @next read only.

sungchoi committed:
Data Structures. Ruby. DoublyLinkedList. Replaces UnderflowError with StandardError in #remove_from_tail.

sungchoi committed:
Data Structures. Ruby. LinkedList. Changes how the module handles empty?
Empty inherits from Node. Empty#empty? returns true. Node#empty? returns false.
Removes type checks for Empty using #empty? instead.

sungchoi committed:
Data Structures. Ruby. BinarySearchTree. First pass. Incomplete.

sungchoi committed:
Data Structures. LinkedList.
Updates #prepend method to LinkedList::Base.
Renames LinkedList::Base#calculate_length to length?.
Adds LinkedList::Node#empty? method.

sungchoi committed:
Data Structures. Ruby. BinarySearchTree. Reorganizes to use classes instead of modules.

sungchoi committed:
Data Structures. Ruby. Queue. Rewrites Queue to use LinkedList instead of DoublyLinkedList.

sungchoi committed:
Data Structures. Ruby. BinaryTreeNode. Updates the argument order for BinaryTreeNode#initialize.
Moves spec/binary_tree/node_spec.rb to spec/binary_tree_node_spec.rb.

sungchoi committed:
Data Structures. Ruby. LinkedList::Base. Adds #prepend tests.

sungchoi committed:
Data Structures. Ruby. BinarySearchTreeNode. Adds #size?(memo) and first test for #size?(memo).

sungchoi committed:
Data Structures. Ruby. Heap. Tree. Cleanup. Removes unused spec/heap/base_spec.rb and spec/tree/node_spec.rb.

sungchoi committed:
Data Structures. Ruby. BinaryTreeNode. BinarySearchTree. BinarySearchTreeNode.
Moves #size?(memo) out of BinarySearchTreeNode and into BinaryTreeNode.
Fills out BinarySearchTree#size? method.

sungchoi committed:
Data Structures. Ruby. BinarySearchTreeNode. Fixes #sorted? method.

sungchoi committed:
Data Structures. Ruby. BinaryTreeNode and BinarySearchTreeNode.
Fixes BinarySearchTreeNode #include?(value) and #insert(value). Needs refinement.
In particular needs to ability to hold a key and a value.

sungchoi committed:
Data Structures. Ruby. BinarySearchTree and BinarySearchTreeNode.
Removes nil and null type checks by adding class methods to BinarySearchTreeNode.

sungchoi committed:
Data Structures. Ruby. BinarySearchTree and BinarySearchTreeNode.
Adds comments.

sungchoi committed:
Data Structures. Ruby. BinaryTreeNode.
Updates #initialize to use options hash instead of additional arguments.

sungchoi committed:
Data Structures. Ruby. BinarySearchTreeNode.
Updates BinarySearchTreeNode to use opts hash.
Uses BinarySearchTreeNode as default for @left and @right.
Fixes #remove(value) method.
Needs refinement.

sungchoi committed:
Data Structures. Ruby. BinarySearchTree.
Updates #remove to assume that BinarySearchTree.remove and BinarySearchTreeNode#remove are not destructive.

sungchoi committed:
Data Structures. Ruby. BinarySearchTreeNode.
Simplifies #insert.
Fixes default assignment in #initialize.
Uses dup in #remove.

sungchoi committed:
Data Structures. Ruby. BinarySearchTreeNode.
Updates #insert to use dup and return a duplicate.

sungchoi committed:
Data Structures. Ruby. BinarySearchTreeNode.
Updates #insert_subtree to use and return duplicate.
Removes private #value= method.

sungchoi committed:
Data Structures. Ruby. BinarySearchTreeNode.
Renames #find_min to #min.
Removes #replace_parent_node method.

sungchoi committed:
Data Structures. Ruby. BinarySearchTree spec.
Fixes miss used variable in #include?(value) returns false when the value the tree is empty test.

sungchoi committed:
Data Structures. Ruby. BinarySearchTree and BinarySearchTreeNode.
Stages changes to make BinarySearchTreeNode expect and to store key value pairs instead of just values.

sungchoi committed:
Data Structures. Ruby. BinarySearchTreeNode and KeyValueBinarySearchTreeNode.
Moves commented out code in BinarySearchTreeNode for a key_value_pair version into its own file and class.

sungchoi committed:
Data Structures. Ruby. BinaryTreeNode.
Removes @parent instance variable.
Adds a parameter to #each to select pre_order, in_order, or post_order.

sungchoi committed:
Data Structures. Ruby. BinarySearchTreeNode.
Fixes references to key back to value.

sungchoi committed:
Data Structures. Ruby. BinarySearchTreeNode.
Fixes references to key_value back to value.
